[AutoFill Debugging] Support HTML as a debug text extraction output format

whsieh opened the pull request:
#### bb1043813cb24812d98028666485be5357f8969a
<pre>
[AutoFill Debugging] Support HTML as a debug text extraction output format
<a href="https://bugs.webkit.org/show_bug.cgi?id=302886">https://bugs.webkit.org/show_bug.cgi?id=302886</a>
<a href="https://rdar.apple.com/165151498">rdar://165151498</a>

Reviewed by Abrar Rahman Protyasha.

Add support for a new text extraction configuration property, `outputFormat`, which encompasses two
values:

`textTree`:     Existing behavior.
`html`:         Format output in a way that&apos;s *similar* to HTML markup (and importantly, can be
                parsed as markup by clients).

See below for more details.

Test: fast/text-extraction/debug-text-extraction-markup.html

* LayoutTests/fast/text-extraction/debug-text-extraction-markup-expected.txt: Added.
* LayoutTests/fast/text-extraction/debug-text-extraction-markup.html: Added.
* Source/WebKit/Shared/TextExtractionToStringConversion.cpp:
(WebKit::escapeStringForHTML):

Add another version of `escapeString`, which is suitable for use in text content in HTML.

(WebKit::TextExtractionAggregator::addResult):
(WebKit::TextExtractionAggregator::useHTMLOutput const):

Add a helper method on `TextExtractionAggregator` indicating whether or not we should use HTML as
the output format.

(WebKit::TextExtractionAggregator::appendToLine):
(WebKit::TextExtractionAggregator::addLineForVersionNumberIfNeeded):
(WebKit::partsForItem):
(WebKit::addPartsForText):
(WebKit::addPartsForItem):

Avoid including rects and event listeners in the text output for now; we may want to add them back
later down the road, as attributes like `rect=[…]` or `eventListeners=[…]`.

(WebKit::addTextRepresentationRecursive):
* Source/WebKit/Shared/TextExtractionToStringConversion.h:
(WebKit::TextExtractionOptions::TextExtractionOptions):
* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(textExtractionOutputFormat):
(-[WKWebView _extractDebugTextWithConfiguration:completionHandler:]):
* Source/WebKit/UIProcess/API/Cocoa/_WKTextExtraction.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKTextExtraction.mm:
(-[_WKTextExtractionConfiguration setOutputFormat:]):
* Tools/TestRunnerShared/UIScriptContext/Bindings/UIScriptController.idl:
* Tools/TestRunnerShared/UIScriptContext/UIScriptController.h:
* Tools/TestRunnerShared/UIScriptContext/UIScriptControllerShared.cpp:
(WTR::toTextExtractionTestOptions):
* Tools/WebKitTestRunner/cocoa/UIScriptControllerCocoa.mm:
(WTR::createTextExtractionConfiguration):

Canonical link: <a href="https://commits.webkit.org/303377@main">https://commits.webkit.org/303377@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/8a4dac8aed3cd527bba692895e13e47073538824

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/132184 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/4677 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/43210 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/139699 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/84108 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/6f2667cf-d826-4667-b8b3-fcb41904acce) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/134054 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/4625 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/4438 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/101040 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/84108 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/3154ad60-6ab6-46ac-bc13-6d8aa7f9067b) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/135130 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/162/builds/3272 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/118389 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/81833 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/94488831-99f4-4373-a0d5-9fbd75b8d44d) 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/154/builds/3161 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac-debug~~](https://ews-build.webkit.org/#/builders/165/builds/1071 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/82919 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/111955 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/36509 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/142346 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/4346 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/37088 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/109414 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/4427 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/161/builds/3769 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/109592 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/27765 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/3298 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/114661 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/57592 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/4400 "Built successfully") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/33048 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/4232 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/67846 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/4491 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/4359 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

aprotyas commented on Source/WebKit/Shared/TextExtractionToStringConversion.cpp:
```diff
@@ -64,6 +64,24 @@ static String escapeString(const String& string)
+    result = makeStringByReplacingAll(result, '\\', "\\\\"_s);
+    result = makeStringByReplacingAll(result, '<', "&lt;"_s);
+    result = makeStringByReplacingAll(result, '>', "&gt;"_s);
+    result = makeStringByReplacingAll(result, '&', "&amp;"_s);
```
This may `s/&gt/&ampgt` right? From the result of the previous two lines?

aprotyas replied:
(I think you want to do the `&`->`&amp` replacement first)

whsieh replied:
🤦🏻‍♂️ Yes, good catch — I'll add a test case for this as well.

aprotyas commented on Source/WebKit/Shared/TextExtractionToStringConversion.cpp:
```diff
@@ -400,100 +447,203 @@ static void addPartsForItem(const TextExtraction::Item& item, std::optional<Node
+                    parts.append(makeString('<', item.nodeName.convertToASCIILowercase(), ' ', makeStringByJoining(attributes, " "_s), '>'));
+
+                if (editableData.isPlainTextOnly)
+                    parts.append("contenteditable=plaintext-only"_s);
```
should `plaintext-only` be quoted? Something like this?
```suggestion
                    parts.append("contenteditable='plaintext-only'"_s);
```

whsieh replied:
Makes sense — that would be consistent with most of the other attributes.

whsieh committed:
[AutoFill Debugging] Support HTML as a debug text extraction output format
https://bugs.webkit.org/show_bug.cgi?id=302886
rdar://165151498

Reviewed by Abrar Rahman Protyasha.

Add support for a new text extraction configuration property, `outputFormat`, which encompasses two
values:

`textTree`:     Existing behavior.
`html`:         Format output in a way that's *similar* to HTML markup (and importantly, can be
                parsed as markup by clients).

See below for more details.

Test: fast/text-extraction/debug-text-extraction-markup.html

* LayoutTests/fast/text-extraction/debug-text-extraction-markup-expected.txt: Added.
* LayoutTests/fast/text-extraction/debug-text-extraction-markup.html: Added.
* Source/WebKit/Shared/TextExtractionToStringConversion.cpp:
(WebKit::escapeStringForHTML):

Add another version of `escapeString`, which is suitable for use in text content in HTML.

(WebKit::TextExtractionAggregator::addResult):
(WebKit::TextExtractionAggregator::useHTMLOutput const):

Add a helper method on `TextExtractionAggregator` indicating whether or not we should use HTML as
the output format.

(WebKit::TextExtractionAggregator::appendToLine):
(WebKit::TextExtractionAggregator::addLineForVersionNumberIfNeeded):
(WebKit::partsForItem):
(WebKit::addPartsForText):
(WebKit::addPartsForItem):

Avoid including rects and event listeners in the text output for now; we may want to add them back
later down the road, as attributes like `rect=[…]` or `eventListeners=[…]`.

(WebKit::addTextRepresentationRecursive):
* Source/WebKit/Shared/TextExtractionToStringConversion.h:
(WebKit::TextExtractionOptions::TextExtractionOptions):
* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(textExtractionOutputFormat):
(-[WKWebView _extractDebugTextWithConfiguration:completionHandler:]):
* Source/WebKit/UIProcess/API/Cocoa/_WKTextExtraction.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKTextExtraction.mm:
(-[_WKTextExtractionConfiguration setOutputFormat:]):
* Tools/TestRunnerShared/UIScriptContext/Bindings/UIScriptController.idl:
* Tools/TestRunnerShared/UIScriptContext/UIScriptController.h:
* Tools/TestRunnerShared/UIScriptContext/UIScriptControllerShared.cpp:
(WTR::toTextExtractionTestOptions):
* Tools/WebKitTestRunner/cocoa/UIScriptControllerCocoa.mm:
(WTR::createTextExtractionConfiguration):

Canonical link: https://commits.webkit.org/303377@main
